### ISSUE TITLE
Why the same training code running on different GPUs (V100 vs P100) brings a significiant difference in outputs and thus train loss / model accuracy?

### ISSUE THREAD

**Issue by lmzhhh** [pytorch/pytorch#58471]
## Why running the same training code on different GPUs (V100 vs P100) brings a significiant difference in outputs ($outputs=model(inputs)$) and thus different train loss / model accuracy?

I run a resnet50 model on cifar10 dataset for only one mini-batch and print out the parameters, inputs, and outputs.

I use the script below for running on a single-GPU. And execute this script on V100 GPU and P100 GPU individually.

```python
CUDA_VISIBLE_DEVICES=0 python -u main.py -e 1 --batch-size 64 -m resnet50 --datatype float32
CUDA_VISIBLE_DEVICES=0 python -u main.py -e 1 --batch-size 64 -m resnet50 --datatype float64
```

I have refered to [https://pytorch.org/docs/stable/notes/randomness.html](https://pytorch.org/docs/stable/notes/randomness.html) for determinastic and reproducible results.
* The results (including train loss, outputs) of the same GPU can be reproduced deterministically. The dumped parameters, inputs, outputs, gradients are all bitwise-identical. (we print out 20 decimal digits of each number, by setting *torch.set_printoptions(profile="full")* )

**However, the results on V100 and P100 (both in float32 and float64 precision) are different.** Specifically, take the first mini-batch for example, their parameters (model.named_parameters()), inputs (batched images) are exactly bitwise-identical, but the outputs (outputs = model(inputs)) are totally different. 
According to IEEE 754 standard, the floating-point operation in float32 should have 6 significant decimal digits precision at least. While the outputs of V100 and P100 only share 2\~5 identical decimal digits. Similarly, the floating-point operation in float64 should have 15 significant decimal digits precision at least. While the outputs of V100 and P100 only share 11\~14 identical decimal digits.

The outputs of the first mini-batch on V100 and P100.

Figure 1. float32
![issue_f32](https://user-images.githubusercontent.com/10664789/118618249-004fce00-b7f6-11eb-9116-81faab715dee.jpg)

Figure 2. float64
![issue_f64](https://user-images.githubusercontent.com/10664789/118618315-1198da80-b7f6-11eb-8130-2371642797ad.jpg)

All the experiments are conducted on an 1 GPU (V100, P100) cloud server (CentOS 7, Driver 450.80.02, Docker 20.10.6, Nvidia Docker 2.5.0) using the official pytorch 180 docker (pytorch/pytorch:1.8.0-cuda11.1-cudnn8-devel).

As the results differ when running on different types of GPUs, I guess the difference may come from the different order of floating-point operations. Which kind of reproducibility is guaranteed by PyTorch?
I am so confused and ask for help here. 

For your convenience, all the code and logs are attached. [resnet50_cifar10_reproduce.tar.gz](https://drive.google.com/file/d/1Sbc767DcM566RI66lTq_2qpCIE2jq2yG/view?usp=sharing)

Thanks!

Mingzhen

**Comment by ailzhang** [comment]
Question for the triage review: do we guarantee determinism across machines?

**Comment by mruberry** [comment]
We do not. From our reproducibility note (https://pytorch.org/docs/master/notes/randomness.html):

> Completely reproducible results are not guaranteed across PyTorch releases, individual commits, or different platforms. Furthermore, results may not be reproducible between CPU and GPU executions, even when using identical seeds.

One possible reason for this: calling into math libraries with different GPUs is probably calling different kernels.
